src/web/lib/braze/forceBrazeMessage.ts

```diff
@@ -0,0 +1,56 @@
+const FORCE_BRAZE_ALLOWLIST = [
+	'preview.gutools.co.uk',
+	'preview.code.dev-gutools.co.uk',
+	'localhost',
+	'm.thegulocal.com',
+];
+
+type Meta = {
+	dataFromBraze: {
+		[key: string]: string;
+	};
+	logImpressionWithBraze: () => void;
+	logButtonClickWithBraze: (id: number) => void;
+};
+
+export const getBrazeMetaFromUrlFragment = (): Meta | null => {
+	if (window.location.hash) {
+		// This is intended for use on development domains for preview purposes.
+		// It won't run in PROD.
+
+		const key = 'force-braze-message';
+
+		const hashString = window.location.hash;
+		const forcedMessage = hashString.slice(
+			hashString.indexOf(`${key}=`) + key.length + 1,
+			hashString.length,
+		);
+
+		if (forcedMessage) {
+			if (!FORCE_BRAZE_ALLOWLIST.includes(window.location.hostname)) {
+				// eslint-disable-next-line no-console
+				console.log(`${key} is not supported on this domain`);
```

Comment:
Does this create a console log statement in production? Ideally (best endeavours) we should aim to have a clean console log for readers

Reply:
It should log to the console - but only if the user has attempted to use a URL fragment including `force-braze-message` (and that will probably only happen intentionally).

I'm not quite sure I was properly checking for that before, so I've made another commit that ensures we will only log to the console if there is a URL fragment and that fragment includes 'force-braze-message'.

Reply:
(I should add - I don't feel particularly strongly about keeping the log as I also like to keep the console clear - I think in this instance though it's unlikely anyone other than Guardian staff will see the log)

Reply:
Is there a particular reason we _need_ to log this information? `console` feels like a tool for debugging.

I think it's generally best practice to remove console log statements from code that is being shipped to production unless there's a strong use case for keeping it.

Reply:
Since this only works in preview, there's a pretty good chance someone in the future will try and drop a message elsewhere and be confused. This is exacerbated by there being loads of reasons in the code why a message might not show up (consent, login, type of subscription), which means that the journey the developer will go on to work out the problem might be a long one. At the time we thought it would be helpful to point out the user error when it occurs, and `console.log` is ideal here because the 'user' (and thus the audience for the message) will be a developer that see console messages.

Are there alternative patterns elsewhere in the codebase for warning developers when they've put things into a known invalid state, to save them having to go on a debugging journey to (re)discover that knowledge?

Reply:
I have mixed feelings about this console log, I am unsure we really want to log this on prod? Its not an error, nor something we should communicate to the user, so it just adds noise to the client IMO.

Reply:
FWIW, I don't hate this considering it's only there for developers who have tried to force something specifically. Maybe you want to use https://github.com/guardian/libs/blob/main/src/logger.README.md with your Team Colours?

Reply:
We’re going to merge this for now given that:

- It will only log if the user is specifically using a URL hash containing `force-braze-message` and therefore shouldn’t impact other teams or our readers.
- It’s a pre-existing log which we’re moving to a new file.

`libs/logger` seems like a good solution but if we were to use it, we would probably want to implement it across all of our logs rather than just this one- so we will look into that in a future PR.